File: Python/DesktopApp/downloader.py
from stage import Stage, Runnable
import os
import logging
import requests
from PyQt5.QtCore import pyqtSlot

class DownloadStage(Stage):
    """Downloads datasets from online sources."""
    def __init__(self, configHandler):
        Stage.__init__(self, configHandler)
        self.downloader = self.Downloader(configHandler)

    @pyqtSlot()
    def download(self):
        self.threadpool.start(self.downloader)

    class Downloader(Runnable):
        def __init__(self, configHandler):
            Runnable.__init__(self, configHandler)
            self.filenameRelPath = "./" + configHandler.getTgzFilename()
            self.datasetType = configHandler.getDatasetType()

        @pyqtSlot()
        def run(self):
            logging.info('Checking if %s already exists', self.filenameRelPath)
            if os.path.isfile(self.filenameRelPath):
                logging.info('%s already exists', self.filenameRelPath)
                logging.info('Checking if %s was downloaded properly', self.filenameRelPath)
                
                if os.path.getsize(self.filenameRelPath) == self.expected_size:
                    logging.info('%s was downloaded properly', self.filenameRelPath)
                    self.signals.attemptUpdateProBarValue.emit(self.getTgzSize())
                    self.signals.attemptUpdateText.emit("Image download complete")
                    self.signals.finished.emit()
                else:
                    logging.warning('%s was NOT downloaded properly', self.filenameRelPath)
                    logging.info('Removing %s', self.filenameRelPath)
                    os.remove(self.filenameRelPath)
                    logging.info('Successfully removed %s', self.filenameRelPath)
                    self.downloadDataset()
            else:
                logging.info('%s does not exist', self.filenameRelPath)
                self.downloadDataset()

        def downloadDataset(self):
            logging.info('Downloading dataset from %s', self.configHandler.getUrl())

            self.signals.attemptUpdateText.emit("Downloading images")
            self.signals.attemptUpdateProBarBounds.emit(0, self.getTgzMax())
            self.signals.attemptUpdateProBarValue.emit(0)

            with requests.get(self.configHandler.getUrl(), stream=True) as r:
                r.raise_for_status() # Raise error if something goes wrong with connection
                with open(self.filenameRelPath, 'wb') as f:
                    for chunk in r.iter_content(chunk_size=8192):
                        if chunk: # filter out keep-alive new chunks
                            self.signals.attemptUpdateProBarValue.emit(self.getTgzSize())
                            f.write(chunk)
                        
            self.run() #######

        def getTgzSize(self):
            """Calculates the size of the TGZ file."""
            if self.datasetType == 'full_set':
                # Dividing by 100 because the expected size of this TGZ is larger than QProgressBar accepts
                return int(os.path.getsize(self.filenameRelPath) / 100)
            elif self.datasetType == 'subset':
                return os.path.getsize(self.filenameRelPath)
            else:
                raise ValueError('Value must be one of the keys in SOURCE_URL')

        def getTgzMax(self):
            """Calculates the size of the TGZ file max."""
            if self.datasetType == 'full_set':
                # Dividing by 100 because the expected size of this TGZ is larger than QProgressBar accepts
                return int(self.expected_size / 100)
            elif self.datasetType== 'subset':
                return self.expected_size
            else:
                raise ValueError('Value must be one of the keys in SOURCE_URL')

File: Go/CxrClassify/downloadStage/downloader.go
package downloadStage

import (
	"CxrClassify/configHandler"
	"CxrClassify/databaseHandler"
	"CxrClassify/runnable"
	"fmt"
	"log"
	"time"

	"io"
	"net/http"
	"os"
)

type Downloader struct {
	runnable.Runnable

	_ func() `constructor:"init"`

	FilenameAbsPath string
	DatasetType     string
}

func (d *Downloader) init() {

}

func (d *Downloader) Setup(configHandler *configHandler.ConfigHandler, databaseHandler *databaseHandler.DatabaseHandler) {
	d.FilenameAbsPath = configHandler.GetTgzFilePath()
	d.DatasetType = configHandler.GetDatasetType()
	d.SetupRunnable(configHandler, databaseHandler)
}

func (d Downloader) Run() {
	k, err := os.OpenFile("testlogfile", os.O_RDWR|os.O_CREATE|os.O_APPEND, 0666)
	if err != nil {
		log.Fatalf("error opening file: %v", err)
	}
	defer k.Close()
	log.SetOutput(k)

	log.Println("Checking if {} already exists")
	info, err := os.Stat(d.FilenameAbsPath)
	if err == nil && !info.IsDir() {
		log.Println("{} already exists")
		log.Println("Checking if {} was downloaded properly")
		if info.Size() == int64(d.Expected_size) {
			log.Println("{} was downloaded properly")
			d.AttemptUpdateText("Image download complete")
			d.Finished()
		} else {
			log.Println("{} was not downloaded properly")
			log.Println("Removing {}")
			e := os.Remove(d.FilenameAbsPath)
			if e != nil {
				log.Println("FAILED")
			}
			log.Println("Successfully removed {}")
		}
	} else {
		log.Println("{} does not exist")

		log.Printf("Downloading dataset from: %s", d.ConfigHandler.GetUrl())

		d.AttemptUpdateText("Downloading images")
		d.AttemptUpdateProBarBounds(0, int(d.getTgzMax()))

		go d.download()

		go func() {

			exists := false
			for !exists {
				_, error := os.Stat("NLMCXR_subset_dataset.tgz")

				// check if error is "file not exists"
				if os.IsNotExist(error) {
					log.Printf("file does not exist\n")
				} else {
					log.Printf("file exist\n")
					exists = true
				}
			}

			var downloaded int64 = -1
			for downloaded < d.getTgzMax() {
				downloaded = d.getTgzSize()
				d.AttemptUpdateProBarValue(int(downloaded))
				time.Sleep(time.Second)
			}

			d.AttemptUpdateText("Image download complete")
			d.Finished()
		}()
	}
}

func (d Downloader) download() error {
	// Create the file
	out, err := os.Create(d.FilenameAbsPath)
	if err != nil {
		return err
	}
	defer out.Close()

	// Get the data
	resp, err := http.Get(d.ConfigHandler.GetUrl())
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	// Check server response
	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("unexpected error code")
	}

	_, err = io.Copy(out, resp.Body)
	if err != nil {
		return err
	}

	return nil
}

func (d Downloader) getTgzSize() int64 {
	fi, err := os.Stat(d.FilenameAbsPath)
	if err != nil {
		log.Println("5")
	}

	if d.DatasetType == "full_set" {
		return int64(fi.Size() / 100)
	} else if d.DatasetType == "subset" {
		return fi.Size()
	} else {
		return 0
	}
}

func (d Downloader) getTgzMax() int64 {
	if d.DatasetType == "full_set" {
		return int64(d.Expected_size / 100)
	} else if d.DatasetType == "subset" {
		return int64(d.Expected_size)
	} else {
		return 0
	}
}
